### ISSUE TITLE
only save a certain number of npm-debug.log files

### ISSUE THREAD

**Issue by othiym23** [npm/npm#7614]
This depends on #6744, and is a followup.

In order to better support `npm report`, while at the same time keeping debug logs from filling up the user's cache / disk / solar system, a cap should be put on the number of debug logs in cache, preferably as a LIFO queue. My strawman suggestion for the number of logs is 20, although if we kept track of the size of the logs, it would also be nice to have a cap on the amount of disk space used by logs as well.
- [ ] only save `x` debug logs (default: 20)
- [ ] only use `y` MiB of disk space for debug logs
- [ ] make the sizes of `x` and `y` configurable
- [ ] expose a way to enumerate and browse the debug logs from the CLI
- [ ] make it easy to attach a specific crash log to [`npm report`](https://github.com/npm/npm/issues/5214)

**Comment by smikes** [comment]
Also ref to #6766 , which has some cleanup logic that may be salvageable.

**Comment by othiym23** [comment]
This is something that's completely within the scope of the larger audit of the user experience that the CLI team would like to in the not-too-distant future (realistically, probably towards the end of 2016).

**Comment by jsdevel** [comment]
@othiym23 is it acceptable then to submit a PR to add a flag in `.npmrc` to disable `npm-debug.log`?

**Comment by othiym23** [comment]
@jsdevel No, the team probably wouldn't merge such a change. From a supportability point of view, that's a dangerous option to support, because if people configure that out of habit, the team (and anyone else trying to troubleshoot) loses their single most valuable form of postmortem debugging.

**Comment by jsdevel** [comment]
Right, but shouldn't that be something the user can take control of?  You can always enable `npm-debug.log` if you need it for support.

**Comment by jsdevel** [comment]
I imagine it being enabled by default.

**Comment by othiym23** [comment]
> shouldn't that be something the user can take control of?

I know this is a contentious position, but in this case, I believe the answer is "no." It genuinely isn't in the best interests of npm's maintainers to add features that make our jobs harder.

I wouldn't object to there being a way on a per-install basis to explicitly disable leaving an `npm-debug.log`, but only explicitly via the command line.

**Comment by jsdevel** [comment]
I guess I just don't see the point.

`console.warn('Errors were found, but you\'ve disabled npm-debug.log.  Re-enable this in .npmrc to see the errors that occurred.');`

To me that feels very reasonable, and wouldn't offend any users who decide to disable `npm-debug.log`.

**Comment by tarikjn** [comment]
Would it make sense at least to be able to redirect npm-debug.log to stderr? This would at least make npm play nicer with Docker.

**Comment by othiym23** [comment]
@tarikjn All npm log output already goes to `stderr` – are you talking about dumping `npm-debug.log` to `stderr` on crash instead of to a file? In practice, `npm --loglevel=silly` will give you a strict superset of what goes into `npm-debug.log`.

**Comment by tarikjn** [comment]
I wasn't aware of that option actually. In case of crash, are there extra logs that are written to the file after npm stops outputting to stderr?
